src/caffe/layers/memory_data_layer.cpp

```diff
@@ -62,13 +89,30 @@ void MemoryDataLayer<Dtype>::Reset(Dtype* data, Dtype* labels, int n) {
 }
 
 template <typename Dtype>
+void MemoryDataLayer<Dtype>::ChangeBatchSize(int new_size) {
+  CHECK(!has_new_data_) <<
+      "Can't change batch_size before all data haven't been consumed"
+      << " by the upper layers";
+  batch_size_ = new_size;
+  added_data_.Reshape(batch_size_, channels_, height_, width_);
```

Comment:
ChangeBatchSize should be just a setter, don't need to reshape anything else. That will happen in the corresponding methods.